### ISSUE TITLE
Code snippet makes the Semantic highlighting very slow

### ISSUE THREAD

**Issue by Goldensteev** [microsoft/pylance-release#2366]
<h3> Environment data </h3>

-   Language Server version: 2022.2.1
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 3.9.1
-   python.analysis.indexing: null
-   python.analysis.typeCheckingMode: off

<h3> Expected behaviour </h3>

Expect Pylance to hightlight my code in seconds as it's been doing for the past few weeks with even longer code snippets

<h3> Actual behaviour </h3>

Semantic Highlighting has been very slow or stuck

<h3> Logs </h3>

Python Language Server Log
```
[Info  - 16:23:51] (13464) Pylance language server 2022.2.1 (pyright 618f54ed) starting
[Info  - 16:23:51] (13464) Server root directory: c:\Users\StevenFletcher\.vscode\extensions\ms-python.vscode-pylance-2022.2.1\dist
[Info  - 16:23:51] (13464) No configuration file found.
[Info  - 16:23:51] (13464) No pyproject.toml file found.
[Info  - 16:23:51] (13464) Setting pythonPath for service "cartographie-ecosysteme-snds": "c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\base39\Scripts\python.exe"
[Warn  - 16:23:51] (13464) stubPath c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\typings is not a valid directory.
[Info  - 16:23:51] (13464) Assuming Python version 3.9
[Info  - 16:23:51] (13464) Assuming Python platform Windows
[Info  - 16:23:52] (13464) Searching for source files
[Info  - 16:23:52] (13464) Auto-excluding c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\base39
[Info  - 16:23:52] (13464) Found 18 source files
[Info  - 16:23:52] (13464) Background analysis(1) root directory: c:\Users\StevenFletcher\.vscode\extensions\ms-python.vscode-pylance-2022.2.1\dist
[Info  - 16:23:52] (13464) Background analysis(1) started
[Info  - 16:23:53] (13464) Indexer background runner(2) root directory: c:\Users\StevenFletcher\.vscode\extensions\ms-python.vscode-pylance-2022.2.1\dist (index)
[Info  - 16:23:53] (13464) Indexing(2) started
[Info  - 16:24:00] (13464) [IDX(2)] Long operation: index execution environment c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds (5671ms)
[Info  - 16:24:00] (13464) [IDX(2)] Long operation: index libraries c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds (6925ms)
[Info  - 16:24:00] (13464) Indexer done(2). indexed 572 files
[Info  - 16:27:48] (13464) [BG(1)] Long operation: getSemanticTokens full at c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\dashboard.py (235523ms)
```

<h3> Code Snippet / Additional information </h3>

slowdown coming from this local import 

```python
import dataprep
```

**Comment by erictraut** [comment]
My, that's quite an expression you have!

Performing type analysis for such a deeply nested expression is computationally expensive. I can look at ways we might be able to speed up this case, but I'll need additional repro steps. This expression is referencing a bunch of symbols whose origin are not clear. For example, I don't know where `app` or `dbc` or `dash_table` come from. Would you be willing to produce a self-contained minimal repro?

If you're looking for a workaround, you can break up the expression into separate expressions that are saved to intermediate variables.

**Comment by Dr-Irv** [comment]
I have a large file (almost 3800 lines) where 2022.2.1 took almost 5 minutes to check the file (as reported by the output log)
Version 2022.2.0 took 4 seconds on the same file.

@erictraut - something funky has happened with the 2022.2.1 release

**Comment by erictraut** [comment]
@Dr-Irv, the issue you're seeing is probably unrelated to this bug report. If you can provide more details (preferably in a separate bug report), that would be appreciated. I have a theory about why you might be seeing this with 2022.2.1, but I'll need additional details to help confirm. I suspect it's this: https://github.com/microsoft/pyright/issues/3008.

**Comment by Dr-Irv** [comment]
> @Dr-Irv, the issue you're seeing is probably unrelated to this bug report. If you can provide more details (preferably in a separate bug report), that would be appreciated. I have a theory about why you might be seeing this with 2022.2.1, but I'll need additional details to help confirm. I suspect it's this: [microsoft/pyright#3008](https://github.com/microsoft/pyright/issues/3008).

I can open up a new issue but not sure what to provide.  I can't share the code.   I just know the performance got significantly worse, and we are seeing this on many files.  The output log from the server shows the analysis time grew.  Tell me what you need and I can open a new issue.

**Comment by erictraut** [comment]
Ah, I was hoping you'd be able to provide a repro case that I could look into. If it's not too much to ask, could you see whether the problem started happening between Pyright 1.1.218 and 1.1.219?

**Comment by Dr-Irv** [comment]
> Ah, I was hoping you'd be able to provide a repro case that I could look into. If it's not too much to ask, could you see whether the problem started happening between Pyright 1.1.218 and 1.1.219?

How?  I only use it within VS Code.  How can I specify the pyright version with VS Code?  And if it is command-line, how do I do it so that it uses the VS Code provided type stubs?  You have to give me more precise instructions on how to do that test.

**Comment by erictraut** [comment]
@Dr-Irv, I found a different repro that confirmed my theory. I have a fix in place, and it will be included in the next release. As I mentioned, I don't think your performance slow-down is related to the issue in this particular bug report.

**Comment by Dr-Irv** [comment]
> @Dr-Irv, I found a different repro that confirmed my theory. I have a fix in place, and it will be included in the next release. As I mentioned, I don't think your performance slow-down is related to the issue in this particular bug report.

So we have to wait 5 days to get the release?  Should we go back to 2022.2.0 to avoid this?  Or is there some other workaround?  It's killing our developer productivity right now.

**Comment by Goldensteev** [comment]
So here is a [repo](https://github.com/Goldensteev/Issue2366-pylance) with the original file and a modified file which still has nested expressions. Funnily enough the modified file runs way faster so something i removed is causing a massive slowdown. 

test.py
```
[Info  - 11:53:53] (14316) Pylance language server 2022.2.1 (pyright 618f54ed) starting
[Info  - 11:53:53] (14316) Server root directory: c:\Users\StevenFletcher\.vscode\extensions\ms-python.vscode-pylance-2022.2.1\dist
[Info  - 11:53:53] (14316) No configuration file found.
[Info  - 11:53:53] (14316) No pyproject.toml file found.
[Info  - 11:53:53] (14316) Setting pythonPath for service "cartographie-ecosysteme-snds": "c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\base39\Scripts\python.exe"
[Warn  - 11:53:53] (14316) stubPath c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\typings is not a valid directory.
[Info  - 11:53:53] (14316) Assuming Python version 3.9
[Info  - 11:53:53] (14316) Assuming Python platform Windows
[Info  - 11:53:54] (14316) Searching for source files
[Info  - 11:53:54] (14316) Auto-excluding c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\base39
[Info  - 11:53:54] (14316) Found 10 source files
[Info  - 11:53:54] (14316) Background analysis(1) root directory: c:\Users\StevenFletcher\.vscode\extensions\ms-python.vscode-pylance-2022.2.1\dist
[Info  - 11:53:54] (14316) Background analysis(1) started
[Info  - 11:53:54] (14316) Indexer background runner(2) root directory: c:\Users\StevenFletcher\.vscode\extensions\ms-python.vscode-pylance-2022.2.1\dist (index)
[Info  - 11:53:54] (14316) Indexing(2) started
[Info  - 11:53:58] (14316) [BG(1)] Long operation: checking: c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\test.py (3202ms)
[Info  - 11:53:58] (14316) [BG(1)] Long operation: analyzing: c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\test.py (3599ms)
[Info  - 11:54:06] (14316) [IDX(2)] Long operation: index execution environment c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds (10251ms)
[Info  - 11:54:06] (14316) [IDX(2)] Long operation: index libraries c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds (11408ms)
[Info  - 11:54:06] (14316) Indexer done(2). indexed 572 files

```
 vs dashboard.py
```
[Info  - 11:18:37] (32228) Pylance language server 2022.2.1 (pyright 618f54ed) starting
[Info  - 11:18:37] (32228) Server root directory: c:\Users\StevenFletcher\.vscode\extensions\ms-python.vscode-pylance-2022.2.1\dist
[Info  - 11:18:37] (32228) No configuration file found.
[Info  - 11:18:37] (32228) No pyproject.toml file found.
[Info  - 11:18:37] (32228) Setting pythonPath for service "cartographie-ecosysteme-snds": "c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\base39\Scripts\python.exe"
[Warn  - 11:18:37] (32228) stubPath c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\typings is not a valid directory.
[Info  - 11:18:37] (32228) Assuming Python version 3.9
[Info  - 11:18:37] (32228) Assuming Python platform Windows
[Info  - 11:18:37] (32228) Searching for source files
[Info  - 11:18:37] (32228) Auto-excluding c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\base39
[Info  - 11:18:37] (32228) Found 10 source files
[Info  - 11:18:37] (32228) Background analysis(1) root directory: c:\Users\StevenFletcher\.vscode\extensions\ms-python.vscode-pylance-2022.2.1\dist
[Info  - 11:18:37] (32228) Background analysis(1) started
[Info  - 11:18:38] (32228) Indexer background runner(2) root directory: c:\Users\StevenFletcher\.vscode\extensions\ms-python.vscode-pylance-2022.2.1\dist (index)
[Info  - 11:18:38] (32228) Indexing(2) started
[Info  - 11:18:44] (32228) [IDX(2)] Long operation: index execution environment c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds (4639ms)
[Info  - 11:18:44] (32228) [IDX(2)] Long operation: index libraries c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds (6109ms)
[Info  - 11:18:44] (32228) Indexer done(2). indexed 572 files
[Info  - 11:49:20] (32228) [BG(1)] Long operation: checking: c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\dashboard.py (1841617ms)
[Info  - 11:49:20] (32228) [BG(1)] Long operation: analyzing: c:\Users\StevenFletcher\Documents\workspace\cartographie-ecosysteme-snds\dashboard.py (1842110ms)

```

**Comment by erictraut** [comment]
@Goldensteev, thanks for preparing the repro. Unfortunately, I'm not able to reproduce the problem you're seeing. The performance looks fine to me (nearly instantaneous semantic highlighting) when using the latest version of pylance. So there must be some additional context I need.

I was able to figure out all of the sources of imports except for the package called `scripts`. That import could be contributing to the slowdown. Could you tell me where `scripts` comes from? I didn't see any obvious candidate packages called `scripts` in the pypi directory. Perhaps it's a local source file that you didn't include in your repro?

Do you have any other non-default "python.analyzer" settings configured in your workspace?

**Comment by ssssshong** [comment]
Same issue here! Semantic highlighting, code completion, parameter suggestions, etc., get super-slowed with 100% cpu usage in Pylance version 2022.2.1. I downgrade to 2022.2.0 and it seems to work fine.

**Comment by erictraut** [comment]
I'm still not able to repro the slow behavior with the code you provided in your repo. The `import dataprep` statement was commented out, so I uncommented it. Is there anything else I need to do to repro the problem? What I see is about a 3 second lag for semantic tokens the first time I open the file. That's reasonable given the complexity of that file and the files it imports. After it is opened the first time, semantic tokens appear immediately when I open the file subsequent times in the same session. It sounds like you're seeing something different?

**Comment by Goldensteev** [comment]
Okay so i have pushed on the repo my .vscode folder which hopefully should make you be able to replicate the Issue i am having as i can replicate it in the repro.
Yes i am seeing something different indeed. When i first load the file with the commented import, i get a fast Highlighting process as per the image 
![image](https://user-images.githubusercontent.com/14848329/154034141-bd197a98-f88e-40e2-a6b1-f1712d645036.png)
 but as soon as i uncomment the import, i get what seems to be an infinite analysis as per image 
![image](https://user-images.githubusercontent.com/14848329/154034322-f373dcd3-fd6e-42fd-ac42-5c62b3df4e2d.png)
![image](https://user-images.githubusercontent.com/14848329/154034348-bf81e6b9-9277-4d82-82f3-2ede27de3e3b.png)
5 mins and file is still not highlighted properly and all subsequent files will not be highlighted as it's stuck on this one
![image](https://user-images.githubusercontent.com/14848329/154034459-01d41936-5436-4945-90a8-0effd85eb530.png)

Also as mentioned, the test file which is a modified dashboard file does work, so it's something specific to the dashboard file and that import.

Hopefully with all this you are able to pinpoint the Issue and replicate it locally!

**Comment by Dr-Irv** [comment]
Here's another sample.  Just pull the pandas repo at https://github.com/pandas-dev/pandas .  Open up `pandas/core/series.py`.  I see the following in the log:
```
[Info  - 8:08:09 AM] (18468) [BG(1)] Long operation: getSemanticTokens full at c:\Code\pandas_dev\mypytest\pandas\core\series.py (16631ms)
[Info  - 8:08:09 AM] (18468) Emptying type cache to avoid heap overflow. Heap size used: 1604MB
[Info  - 8:08:26 AM] (18468) [BG(1)] Long operation: checking: c:\Code\pandas_dev\mypytest\pandas\core\series.py (16387ms)
[Info  - 8:08:26 AM] (18468) [BG(1)] Long operation: analyzing: c:\Code\pandas_dev\mypytest\pandas\core\series.py (16387ms)
[Info  - 8:08:33 AM] (18468) Emptying type cache to avoid heap overflow. Heap size used: 2055MB
[Info  - 8:08:43 AM] (18468) Emptying type cache to avoid heap overflow. Heap size used: 2422MB
[Warn  - 8:08:46 AM] (18468) Workspace indexing has hit its upper limit: 2000 files
```

**Comment by erictraut** [comment]
@Dr-Irv, we understand the issue you're seeing, and it should be addressed in this week's Pylance release. 

@Goldensteev, I'm still not seeing the same behavior you're reporting. Here are the steps I'm using to try to reproduce the issue. Please let me know if I have these correct.

1. Clone the repo [here](https://github.com/Goldensteev/Issue2366-pylance).
2. Make sure Pylance v2022.2.1 is installed.
3. Open the root directory from the cloned repo and open the file `dashboard.py`
4. Uncomment the line `# import dataprep`

Here's what I see prior to uncommenting the import:
```
[Info  - 8:25:46 AM] (45055) [BG(1)] Long operation: getSemanticTokens full at /xxx/Issue2366-pylance/dashboard.py (2565ms)
```

Here's what I see after uncommenting the import:
```
[Info  - 8:27:01 AM] (45197) [BG(1)] Long operation: getSemanticTokens full at /xxx/Issue2366-pylance/dashboard.py (5100ms)
```

Both of these times are long, but I'm not seeing anything approaching 5 minutes. Is there anything in the above steps that I missed? Or anything else about your configuration that might be relevant? For example, are you editing remotely over ssh?

Perhaps another member of the Pylance team could try to repro following these steps (@debonte, @heejaechang)?

**Comment by Goldensteev** [comment]
this new patch did not fix my issue unfortunately

**Comment by Dr-Irv** [comment]
> this new patch did not fix my issue unfortunately

The new patch _did_ fix both the issues I reported above https://github.com/microsoft/pylance-release/issues/2366#issuecomment-1036434796  and https://github.com/microsoft/pylance-release/issues/2366#issuecomment-1040253814

**Comment by ssssshong** [comment]
> Same issue here! Semantic highlighting, code completion, parameter suggestions, etc., get super-slowed with 100% cpu usage in Pylance version 2022.2.1. I downgrade to 2022.2.0 and it seems to work fine.

The issues are fixed in the new version, 2022.2.3. Thank you!

**Comment by radu-ion** [comment]
Same thing with me. Pylance 2022.2.0 does semantic highlighting instantly on `import sys`, Pylance 2022.2.3 gets stuck (it does not work for me). I don't have logs but I will stick with 2022.2.0 for a bit longer.
It's not just about semantic highlighting, plain syntax checking fails, e.g. 'a =' followed by a newline does not yield a red squiggle. If I switch to Jedi, it works fine.

Running Windows 10, 64 bit, Python 3.9.9, default settings for all extensions.
![Not working](https://user-images.githubusercontent.com/18706300/155155133-bfd88e51-cad5-417c-9794-88058afd355e.jpg)
![Working](https://user-images.githubusercontent.com/18706300/155155141-de9762f6-f614-4858-a35e-045dab877788.jpg)

**Comment by erictraut** [comment]
@radu-ion, could you post a minimal, self-contained piece of code that reproduces the problem? We won't be able to debug an issue like this from partial screen shots. Thanks!

**Comment by Goldensteev** [comment]
So I was hoping to come here this morning and comment "Hallelujah" but unfortunately it seems the issue is not fixed for me.
I have of course updated to the last version and it is still extremely slow in that one file.
The full project has gone open source so i will share with you the full git which hopefully will help pinpoint the root cause of this issue.

https://gitlab.com/healthdatahub/cartographie-ecosysteme-snds

As stated above, it seems there where 2 separate issues perhaps, the nested expressions, and that import.

in the _dashboard.py_ file, commenting `import lib.data as data` will make the semantic highlighting act normally, but uncommenting it will make the highlight very very slow, and with the new version still.

![image](https://user-images.githubusercontent.com/14848329/156729040-345e3901-e356-4a78-80f2-dae6a3f87f51.png)
first time being with commented lib import and then without.

if there is any additionnal info that can be shared please ask i will give it.

Thanks

**Comment by erictraut** [comment]
@Goldensteev, thanks for posting — and for the repro steps. 

Would you mind opening a new issue? This one has already been fixed and closed. While the symptoms you're seeing are similar, it's clearly a different issue, and I don't want it to get lost.

**Comment by erictraut** [comment]
Never mind, I just realized that this is probably the same issue that you reported at the start of this thread. I'll re-open this issue.

**Comment by erictraut** [comment]
Since this is a core type checking issue, I've created a tracking bug in the pyright repo: https://github.com/microsoft/pyright/issues/3145

**Comment by erictraut** [comment]
@Goldensteev, I was able to repro the perf issue. I found the underlying cause. It will be fixed in the next release. Thanks again for reporting the problem.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2022.3.1, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202231-9-March-2022)
